From 7a8eb8fa355256c4f4854b843b1c53f4ae37398f Mon Sep 17 00:00:00 2001
From: hariprasad <hkelam@marvell.com>
Date: Tue, 7 Jan 2020 16:17:14 +0530
Subject: [PATCH 445/767] octeontx2-pf: Reset HW config for HIGIG/EDSA in
 driver unbind

commit dfc7abe46c1bf3c2c00fa43003f3e11650e21632 from
git@git.assembla.com:cavium/WindRiver.linux.git

Once user enabled HIGIG/EDSA to parse these packets specific
pkind will be alloted. For HIGIG CGX/LMAC registers are configured
to support higig pause frames etc.This patch reset these settings
in driver unbind

Change-Id: I61cb5255e5f3d97c8caa2c86565e4e7e35f43aba
Signed-off-by: hariprasad <hkelam@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/21180
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[Kevin: Just some minor context mods in order to port to linux-yocto]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 .../net/ethernet/marvell/octeontx2/nic/otx2_common.h   |  2 +-
 .../net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c  | 10 +++++++---
 drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c   |  6 +++---
 3 files changed, 11 insertions(+), 7 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_common.h b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_common.h
index 548ec6075bc9..cfcf50c57fce 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_common.h
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_common.h
@@ -751,7 +751,7 @@ int otx2vf_open(struct net_device *netdev);
 int otx2vf_stop(struct net_device *netdev);
 int otx2_set_real_num_queues(struct net_device *netdev,
 			     int tx_queues, int rx_queues);
-int otx2_set_npc_parse_mode(struct otx2_nic *pfvf);
+int otx2_set_npc_parse_mode(struct otx2_nic *pfvf, bool unbind);
 
 /* MCAM filter related APIs */
 void otx2_do_set_rx_mode(struct work_struct *work);
diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
index 7b9cbca6d9df..acd6e690a371 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_ethtool.c
@@ -1311,7 +1311,7 @@ static int otx2_set_phy_mod_type(struct net_device *netdev, bool enable)
 	return rc;
 }
 
-int otx2_set_npc_parse_mode(struct otx2_nic *pfvf)
+int otx2_set_npc_parse_mode(struct otx2_nic *pfvf, bool unbind)
 {
 	struct npc_set_pkind *req;
 	int rc = -EAGAIN;
@@ -1321,7 +1321,9 @@ int otx2_set_npc_parse_mode(struct otx2_nic *pfvf)
 	if (!req)
 		goto end;
 
-	if (OTX2_IS_HIGIG2_ENABLED(pfvf->ethtool_flags))
+	if (unbind)
+		req->mode = OTX2_PRIV_FLAGS_DEFAULT;
+	else if (OTX2_IS_HIGIG2_ENABLED(pfvf->ethtool_flags))
 		req->mode = OTX2_PRIV_FLAGS_HIGIG;
 	else if (OTX2_IS_EDSA_ENABLED(pfvf->ethtool_flags))
 		req->mode = OTX2_PRIV_FLAGS_EDSA;
@@ -1331,11 +1333,13 @@ int otx2_set_npc_parse_mode(struct otx2_nic *pfvf)
 	req->dir  = PKIND_RX;
 
 	/* req AF to change pkind on both the dir */
-	if (req->mode == OTX2_PRIV_FLAGS_HIGIG)
+	if (req->mode == OTX2_PRIV_FLAGS_HIGIG || unbind)
 		req->dir |= PKIND_TX;
 
 	if (!otx2_sync_mbox_msg(&pfvf->mbox))
 		rc = 0;
+	else
+		pfvf->ethtool_flags &= ~req->mode;
 end:
 	otx2_mbox_unlock(&pfvf->mbox);
 	return rc;
diff --git a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
index a62b88c88d32..6a7966770d26 100644
--- a/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
+++ b/drivers/net/ethernet/marvell/octeontx2/nic/otx2_pf.c
@@ -1658,9 +1658,7 @@ int otx2_open(struct net_device *netdev)
 
 	/* Set NPC parsing mode, skip LBKs */
 	if (!is_otx2_lbkvf(pf->pdev)) {
-		err = otx2_set_npc_parse_mode(pf);
-		if (err)
-			goto err_free_cints;
+		otx2_set_npc_parse_mode(pf, false);
 	}
 
 	err = otx2_rxtx_enable(pf, true);
@@ -2668,6 +2666,8 @@ static void otx2_remove(struct pci_dev *pdev)
 	if (pf->flags & OTX2_FLAG_RX_TSTAMP_ENABLED)
 		otx2_config_hw_rx_tstamp(pf, false);
 
+	otx2_set_npc_parse_mode(pf, true);
+
 	/* Disable link notifications */
 	otx2_cgx_config_linkevents(pf, false);
 
-- 
2.31.1

